pkg/rpc/ethereum_client.go

```
@@ -4,66 +4,187 @@ import (
+			if err != nil {
+				logger.Sugar.Errorf("Failed to create HTTP client with API key: %v", err)
+				return nil, errors.NewRPCConnectionFailed("rpc", "NewEthereumClient", httpNodeURL, err)
+			}
+			httpClient = ethclient.NewClient(rpcClient)
+			logger.Sugar.Info("HTTP client with API key created successfully")
+		} else {
+			logger.Sugar.Info("Creating HTTP client without API key")
+			// Standard connection without API key
+			httpClient, err = ethclient.Dial(httpNodeURL)
+			if err != nil {
+				return nil, errors.NewRPCConnectionFailed("rpc", "NewEthereumClient", httpNodeURL, err)
+			}
 		}
 		client.httpClient = httpClient
 	}
 
+	// Establish WebSocket client with API key authentication if provided
+	if wsURL != "" {
+		logger.Sugar.Infof("Attempting WebSocket connection to %s", wsURL)
+		var wsClient *ethclient.Client
+		var err error
+
+		if apiKey != "" {
+			// Create WebSocket connection with custom headers for GCP authentication
+			logger.Sugar.Info("Creating WebSocket connection with X-goog-api-key header")
+			wsClient, err = createWebSocketWithHeaders(wsURL, apiKey)
+			if err != nil {
+				logger.Sugar.Warnf("Failed to establish WebSocket connection with API key header: %v", err)
+				// Try fallback without API key
+				logger.Sugar.Info("Trying standard WebSocket connection as fallback")
+				wsClient, err = ethclient.Dial(wsURL)
+				if err != nil {
+					logger.Sugar.Warnf("Failed to establish fallback WebSocket connection: %v", err)
+				} else {
+					logger.Sugar.Info("WebSocket fallback connection successful")
+					client.wsClient = wsClient
+				}
+			} else {
+				logger.Sugar.Info("WebSocket connection with API key header successful")
+				client.wsClient = wsClient
+			}
+		} else {
+			// Standard WebSocket connection without API key
+			wsClient, err = ethclient.Dial(wsURL)
+			if err != nil {
+				logger.Sugar.Warnf("Failed to establish WebSocket connection: %v", err)
+				// Don't fail completely, HTTP client can still work
```

The client is only created during startup. If someone has configured a websocket address and they fail to connect to it, we should probably return an error. We can still return the client and allow the caller to choose how to handle it.

Imagine someone makes a typo during config, misses the warning log, and then are later confused to see that their indexer isn't performing as fast as they expected (if they even find out)